_posts/wasmer-go-embedding-1.0.md

```diff
@@ -0,0 +1,534 @@
+
+1. There is an `Engine` API; more about this in the next section,
+
+2. Host functions are fully implemented in Go; `cgo` is no longer
+   necessary when declaring host functions. Moreover, host functions
+   no longer receive the famous `context` argument; more on that in a
+   second,
+
+3. Host functions can return multiple-values,
+
+4. Host functions can return any error _à la_ Go,
+
+5. The `ImportObject` API allows to register a collection of
+   “externals”: `Function`, `Memory` etc., i.e. any type that
+   implements the `IntoExtern` interface, to a given namespace (here
+   `math`),
+   
+6. The `Exports` API allows to read any kinds of externals, including
+   multiple memories.
+
+The entire API is more Go idiomatic, and very much simpler to use from
+our perspective.
+
+### A word about host functions
+
+Previously, a host function had to be declared with a `cgo` mapping:
+
+```go
+// extern int32_t sum(void *context, int32_t x, int32_t y);
+import "C"
+
+//export sum
+func sum(context unsafe.Pointer, x int32, y int32) int32 {
+	return x + y
+}
+```
+
+Now, a host function no longer needs `cgo`. That's a big
+user-experience improvement! A host function can be any Go functions
+or Go lambda functions.
+
+Also, a host function can now return an error, which was impossible
+before. It will automatically translate to a WebAssembly trap (with
+the new `Trap` API).
+
+Finally, a host function receive a slice of values and returns another
+slices of values, i.e. it supports multi-values!
+
+```go
+func sum(args []wasmer.Value) ([]wasmer.Value, error) {
+	x := args[0].I32()
+	y := args[1].I32()
+
+	return []wasmer.Value{wasmer.NewI32(x + y)}, nil
+}
+```
+
+Note that a host function manipulates values of kind `Value` to allow
+a more unified API.
+
+We are talking about the host function implementation. To create a
+proper host function, we use the `NewFunction` function. It is also
+possible to create a new host function with an attached environment
+with `NewFunctionWithEnvironment` which expects a function of the
+following form:
+
+```go
+type MyEnvironment struct {
+	shift int32
+}
+
+environment := &MyEnvironment {
+	shift: 42,
+}
+
+hostFunction := wasmer.NewFunction(
+	store,
+
+	// The host function signature.
+	wasmer.NewFunctionType(
+		wasmer.NewValueTypes(wasmer.I32, wasmer.I32), // two i32 params
+		wasmer.NewValueTypes(wasmer.I32), // one i32 result
+	),
+
+	// Our environment!
+	environment,
+
+	// The function implementation.
+	func(environment interface{}, args []wasmer.Value) ([]wasmer.Value, error) {
+		// Cast to our environment type, and do whatever we want!
+		env := environment.(*MyEnvironment)
+		e := env.shift;
+		x := args[0].I32()
+		y := args[1].I32()
+
+		return []wasmer.Value{wasmer.NewI32(e + x + y)}, nil
+	},
+)
+```
+
+The biggest improvement is that the environment can be anything! No
+more restriction due to `cgo`. The user is free to store anything
+needed.
```

Comment:
Technically it should be `Send + Sync` because we may share it across threads but we can deal with that later if that has an impact on users

Reply:
Actually no, it's truely anything. The environment in Go isn't send into Rust! An associated environment index is sent on the Rust side, but the environment stays on the Go side, inside a concurent-locked space. See https://github.com/wasmerio/wasmer-go/blob/011e5004cb286a2a7473387aae193a690f690760/wasmer/function.go#L352-L411.

I had too much issue with environment being sent to Rust, or going through cgo. That's why I've decided to design the new API like this.